syserol/tensor.py

```diff
@@ -127,6 +128,44 @@ def censored_lstsq(A: np.ndarray, B: np.ndarray, uniqueInfo) -> np.ndarray:
     return X.T
 
 
+def sigmoid(x: np.ndarray, P: np.ndarray):
+    """ Basic sigmoidal function """
+    x0, k = P
+    y = 1 / (1 + np.exp(-k*(x-x0)))
+    return y
+
+
+def F(P: np.ndarray, x: np.ndarray, y: np.ndarray):
+    return(sigmoid(x, P) - y)
+
+
+def reverse_p(x: np.ndarray, y: np.ndarray, p_init: np.ndarray):
+    """ Solves for parameter matrix given an output y from our sigmoidal time function 
+    In this case, x is our time vector, y is our solution to the sigmoidal function.
+    """
+    res = sp.optimize.least_squares(F, p_init, args=(x, y))
+    return(res.x)
+
+
+def continue_solve(v: np.ndarray, A_hat: np.ndarray):
+    """ Reverse solves for parameter matrix P_new given current guess for time factor.
+    Updates current iterate of time factor (A_new) using solved parameters. 
+    P_new : params x r matrix
+    A_new : r x n matrix 
+    Note that the first iteration initializes P to be all ones.
+    """
+    rank = A_hat.shape[1]
+    A_new = np.empty(A_hat.shape)
+    P_new = np.empty((4, rank))
+
+    for comp in range(rank):
+        p_est = np.polyfit(v, A_hat[:, comp], 3)
```

Comment:
Since you said the sigmoid had a higher R2X, it sounds like that's working reasonably well. I'd just switch back to that.

Reply:
Seems like it has a _slightly_ higher R2X. But sometimes we get negatives. So there is a tradeoff. Here are the sigmoidal R2X values, for one run (but seem consistent across runs):
0.4241676606692326
0.5073001598024234
0.5706641230936531
0.5781722025811826
0.5857755069186128
0.6183951413933952
0.6867237126389697
0.6927166798657362
0.6700033482041197
-110323.53578568765

Here are the polynomial:
0.4170025806266322
0.4416334897013897
0.5445742119648187
0.4249694556266973
0.47633973748995373
0.5464716601893254
0.5882739537713366
0.5792293111589037
0.45179798124621906
0.6447411913973247

Reply:
Hmm... that's weird because as far as I can tell all your code is deterministic, except for the random number padding in U. What number of components are you using?

Reply:
(You shouldn't be hitting the random number padding.)

Reply:
Oh, sorry. This is outputs for components 1-10 in each.

Reply:
For example, component 10 for polynomial is 0.64 R2X. For the sigmoidal, we got a negative (but it's not always that?). I thought that was weird too, because I don't believe the solves change run to run

Reply:
Ok, 10 components is reeeeeeeeaaaally high for this dataset. Just limit yourself to 8 and go with the sigmoid.

Reply:
Like, we use 2 components in the MSB paper

Reply:
Sounds good.